In [2]:
import pandas as pd
import numpy as np


In [4]:
path = "C:/Users/micho/OneDrive/Desktop/DATABASE/PROJECTS HOUSE/Machine-Learning-Tutorials-Scikit-Learn-main/heart_failure_clinical_records_dataset.csv"

In [5]:
df = pd.read_csv(path)

In [6]:
df

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    75.0        0                       582         0                 20   
1    55.0        0                      7861         0                 38   
2    65.0        0                       146         0                 20   
3    50.0        1                       111         0                 20   
4    65.0        1                       160         1                 20   
..    ...      ...                       ...       ...                ...   
294  62.0        0                        61         1                 38   
295  55.0        0                      1820         0                 38   
296  45.0        0                      2060         1                 60   
297  45.0        0                      2413         0                 38   
298  50.0        0                       196         0                 45   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                      1  265000.00               1.9           130    1   
1                      0  263358.03               1.1           136    1   
2                      0  162000.00               1.3           129    1   
3                      0  210000.00               1.9           137    1   
4                      0  327000.00               2.7           116    0   
..                   ...        ...               ...           ...  ...   
294                    1  155000.00               1.1           143    1   
295                    0  270000.00               1.2           139    0   
296                    0  742000.00               0.8           138    0   
297                    0  140000.00               1.4           140    1   
298                    0  395000.00               1.6           136    1   

     smoking  time  DEATH_EVENT  
0          0     4            1  
1          0     6            1  
2          1     7            1  
3          0     7            1  
4          0     8            1  
..       ...   ...          ...  
294        1   270            0  
295        0   271            0  
296        0   278            0  
297        1   280            0  
298        1   285            0  

[299 rows x 13 columns]

In [7]:
df.columns

Index(['age', 'anaemia', 'creatinine_phosphokinase', 'diabetes',
       'ejection_fraction', 'high_blood_pressure', 'platelets',
       'serum_creatinine', 'serum_sodium', 'sex', 'smoking', 'time',
       'DEATH_EVENT'],
      dtype='object')

In [8]:
df['DEATH_EVENT'].value_counts()

DEATH_EVENT
0    203
1     96
Name: count, dtype: int64

In [9]:
df.isna().sum()

age                         0
anaemia                     0
creatinine_phosphokinase    0
diabetes                    0
ejection_fraction           0
high_blood_pressure         0
platelets                   0
serum_creatinine            0
serum_sodium                0
sex                         0
smoking                     0
time                        0
DEATH_EVENT                 0
dtype: int64

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       299 non-null    float64
 1   anaemia                   299 non-null    int64  
 2   creatinine_phosphokinase  299 non-null    int64  
 3   diabetes                  299 non-null    int64  
 4   ejection_fraction         299 non-null    int64  
 5   high_blood_pressure       299 non-null    int64  
 6   platelets                 299 non-null    float64
 7   serum_creatinine          299 non-null    float64
 8   serum_sodium              299 non-null    int64  
 9   sex                       299 non-null    int64  
 10  smoking                   299 non-null    int64  
 11  time                      299 non-null    int64  
 12  DEATH_EVENT               299 non-null    int64  
dtypes: float64(3), int64(10)
memory usage: 30.5 KB


In [11]:
X = df.drop('DEATH_EVENT', axis = 1)

In [12]:
X

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    75.0        0                       582         0                 20   
1    55.0        0                      7861         0                 38   
2    65.0        0                       146         0                 20   
3    50.0        1                       111         0                 20   
4    65.0        1                       160         1                 20   
..    ...      ...                       ...       ...                ...   
294  62.0        0                        61         1                 38   
295  55.0        0                      1820         0                 38   
296  45.0        0                      2060         1                 60   
297  45.0        0                      2413         0                 38   
298  50.0        0                       196         0                 45   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                      1  265000.00               1.9           130    1   
1                      0  263358.03               1.1           136    1   
2                      0  162000.00               1.3           129    1   
3                      0  210000.00               1.9           137    1   
4                      0  327000.00               2.7           116    0   
..                   ...        ...               ...           ...  ...   
294                    1  155000.00               1.1           143    1   
295                    0  270000.00               1.2           139    0   
296                    0  742000.00               0.8           138    0   
297                    0  140000.00               1.4           140    1   
298                    0  395000.00               1.6           136    1   

     smoking  time  
0          0     4  
1          0     6  
2          1     7  
3          0     7  
4          0     8  
..       ...   ...  
294        1   270  
295        0   271  
296        0   278  
297        1   280  
298        1   285  

[299 rows x 12 columns]

In [13]:
y =  df['DEATH_EVENT']

In [14]:
y

0      1
1      1
2      1
3      1
4      1
      ..
294    0
295    0
296    0
297    0
298    0
Name: DEATH_EVENT, Length: 299, dtype: int64

In [15]:
from sklearn.model_selection import train_test_split

In [16]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size= 0.4, random_state= 5)

In [17]:
from sklearn.tree import DecisionTreeClassifier

In [18]:
DTC = DecisionTreeClassifier()
# The Decision Tree cconstructor carries no parameter besides the default parameters

In [19]:
DTC.fit(X_train, y_train)

DecisionTreeClassifier()

In [20]:
DTC.score(X_train, y_train)

1.0

The model score 100% accuracy score on the training dataset. An overfitting case here

In [21]:
DTC.score(X_test, y_test)

0.775

The model scores 76.6 % on the test dataset(Out of sample accuracy)

In [23]:
# Parameter Tunning

DTC1 = DecisionTreeClassifier(max_depth=8)
DTC1.fit(X_train, y_train)
DTC1.score(X_test, y_test)
print('accuracy score for the train data: ', DTC1.score(X_train, y_train))
print('accuracy score for the test data: ', DTC1.score(X_test, y_test))

# The Decision Tree cconstructor carries a parameter specified for it

accuracy score for the train data:  1.0
accuracy score for the test data:  0.7583333333333333


## Using the GridSearchCV to help choose the best parametr for achieving higher accuracy

In [24]:
parameters = {'max_depth':[3, 4, 5, 6, 7], 'max_leaf_nodes':[2,3,4,5,6]}
DTC2 = DecisionTreeClassifier()


In [25]:
from sklearn.model_selection import GridSearchCV
GS = GridSearchCV(DTC2, parameters)
GS.fit(X_train, y_train)


C:\Users\micho\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\numpy\ma\core.py:2846: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,


GridSearchCV(estimator=DecisionTreeClassifier(),
             param_grid={'max_depth': [3, 4, 5, 6, 7],
                         'max_leaf_nodes': [2, 3, 4, 5, 6]})

In [26]:
print('accuracy of train dataset:', GS.score(X_train, y_train))
print('accuracy of test data:', GS.score(X_test, y_test))

accuracy of train dataset: 0.8547486033519553
accuracy of test data: 0.8333333333333334


In [27]:
GS.best_params_

{'max_depth': 3, 'max_leaf_nodes': 2}

The best parameters from the Grid serach falls on :{'max_depth': 3, 'max_leaf_nodes': 2}

---



In [28]:

DTC3 = DecisionTreeClassifier(max_depth = 3, max_leaf_nodes=2)

DTC3.fit(X_train, y_train)
DTC3.score(X_test, y_test)
print('accuracy score for the train data: ', DTC3.score(X_train, y_train))
print('accuracy score for the test data: ', DTC3.score(X_test, y_test))


accuracy score for the train data:  0.8547486033519553
accuracy score for the test data:  0.8333333333333334


### Just being inquisitive here with parameter

In [29]:

DTC4 = DecisionTreeClassifier(max_depth = 2, max_leaf_nodes=2)

DTC4.fit(X_train, y_train)
DTC4.score(X_test, y_test)
print('accuracy score for the train data: ', DTC4.score(X_train, y_train))
print('accuracy score for the test data: ', DTC4.score(X_test, y_test))


accuracy score for the train data:  0.8547486033519553
accuracy score for the test data:  0.8333333333333334


## Using the Confusion Matrix besides the accuracy score to evaluate the accuracy of the model

In [31]:
from sklearn import tree
from sklearn.metrics import confusion_matrix

In [32]:
y_predict = DTC3.predict(X_test)
y_predict

array([0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1,
       0, 1, 0, 0, 0, 1, 0, 1, 0, 0])

In [33]:
confusion_matrix(y_test, y_predict)

array([[74,  4],
       [16, 26]])

In [34]:
Tp,Fn, Fp,Tn = confusion_matrix(y_test, y_predict).ravel()
Tp,Fn, Fp,Tn

(np.int64(74), np.int64(4), np.int64(16), np.int64(26))

# Using the input variables to predict the output/target In order to which is the Death Event of the whole dataset.

In [35]:
yhat = DTC3.predict(X)
yhat

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [36]:
confusion_matrix(y, yhat)

array([[193,  10],
       [ 36,  60]])

In [37]:
Tp,Fn, Fp,Tn = confusion_matrix(y, yhat).ravel()
Tp,Fn, Fp,Tn

(np.int64(193), np.int64(10), np.int64(36), np.int64(60))

# The model could predict 193 correctly out of 203 No Death Event and 60 correctly out of 96 Death Event

## Author
[Jolayemi Sunday Babatunde](linkedin.com/in/babatunde-jolayemi-a05312275)